src/DataAccess/DoctrineMembershipApplicationPiwikTracker.php

```diff
@@ -4,55 +4,35 @@
 
 namespace WMDE\Fundraising\Frontend\DataAccess;
 
-use Doctrine\ORM\EntityManager;
-use Doctrine\ORM\ORMException;
 use WMDE\Fundraising\Entities\MembershipApplication;
 use WMDE\Fundraising\Frontend\Infrastructure\MembershipApplicationPiwikTracker;
 use WMDE\Fundraising\Frontend\Infrastructure\MembershipApplicationPiwikTrackingException;
+use WMDE\Fundraising\Store\MembershipApplicationRepository;
+use WMDE\Fundraising\Store\MembershipApplicationRepositoryException;
 
 /**
  * @licence GNU GPL v2+
  * @author Jeroen De Dauw < jeroendedauw@gmail.com >
  */
 class DoctrineMembershipApplicationPiwikTracker implements MembershipApplicationPiwikTracker {
 
-	private $entityManager;
+	private $applicationModifier;
 
-	public function __construct( EntityManager $entityManager ) {
-		$this->entityManager = $entityManager;
+	public function __construct( MembershipApplicationRepository $applicationModifier ) {
+		$this->applicationModifier = $applicationModifier;
 	}
 
 	public function trackApplication( int $applicationId, string $trackingString ) {
-		$application = $this->getApplicationById( $applicationId );
-
-		$application->setTracking( $trackingString );
-
-		$this->persistApplication( $application );
-	}
-
-	private function getApplicationById( int $applicationId ): MembershipApplication {
-		try {
-			$application = $this->entityManager->find( MembershipApplication::class, $applicationId );
-		}
-		catch ( ORMException $ex ) {
-			// TODO: might want to log failure here
-			throw new MembershipApplicationPiwikTrackingException( 'Membership application could not be accessed' );
-		}
-
-		if ( $application === null ) {
-			throw new MembershipApplicationPiwikTrackingException( 'Membership application does not exist' );
-		}
-
-		return $application;
-	}
-
-	private function persistApplication( MembershipApplication $application ) {
 		try {
-			$this->entityManager->persist( $application );
-			$this->entityManager->flush();
+			$this->applicationModifier->modifyApplication(
+				$applicationId,
+				function( MembershipApplication $application ) use ( $trackingString ) {
+					$application->setTracking( $trackingString );
+				}
+			);
```

Comment:
Should also call `$this->applicationModifier->flush()` the way it is currently implemented